File: 2019/day-01/1.cpp
#include<iostream>
#include<string>
#include<cmath>

std::string g_mass;
int g_fuel1, g_fuel2;

int massToFuel(int mass) {
    int fuel = floor(mass / 3.0) - 2;

    return (fuel > 0)? fuel : 0;
}

int smartMassToFuel(int mass) {
    int fuel = 0;
    int addFuel = massToFuel(mass);

    while (addFuel > 0) {
        fuel += addFuel;
        addFuel = massToFuel(addFuel);
    }

    return fuel;
}

int main() {
    freopen("../../inputs/2019/01.txt", "r", stdin);

    while (std::cin >> g_mass) {
        g_fuel1 += massToFuel(std::stoi(g_mass));
        g_fuel2 += smartMassToFuel(std::stoi(g_mass));
    }

    std::cout << "Part 1: " << g_fuel1 << "\n";
    std::cout << "Part 2: " << g_fuel2 << "\n";

    return 0;
}

File: 2019/day-01/1.py
inputFile = "../../inputs/2019/01.txt"

def readFile():
    file = open(inputFile, "r")
    if (file.mode != 'r'): exit()
    lines = file.readlines()
    file.close()
    return lines

def massToFuel(mass):
    return mass // 3 - 2

def smartMassToFuel(mass):
    fuel = 0
    add = massToFuel(mass)
    
    while (add > 0):
        fuel += add
        add = massToFuel(add)
    
    return fuel

def main():
    fuelA, fuelB = 0, 0

    lines = readFile()

    for line in lines:
        fuelA += massToFuel(int(line))
        fuelB += smartMassToFuel(int(line))

    print("Part 1: " + str(fuelA))
    print("Part 2: " + str(fuelB))

if __name__ == '__main__': main()